server: re-implement table name validity filtering

abooij opened the pull request:
<!-- Thank you for submitting this PR! :) -->
<!-- Provide a general summary of your changes in the Title above ^, end with (close #<issue-no>) or (fix #<issue-no>) -->

### Description
<!-- The title might not be enough to convey how this change affects the user. -->
<!-- Describe the changes from a user's perspective -->

#6116 was based on a rebase of [be84310a4c9f4757575c8f59241d602c1c78ce6c](https://github.com/hasura/graphql-engine/pull/5797/commits/be84310a4c9f4757575c8f59241d602c1c78ce6c), which is part of PR #5797, which is based on b2561a719bb586a74dc40515a6f4637c79b91f03.  However, since branching, the only file touched by #6116, namely `server/src-lib/Hasura/GraphQL/Schema.hs`, underwent a number of changes, and #6116 did not incorporate those of #5719 entirely. This PR attempts to fix that, in particular filtering tables and functions on having a valid GraphQL name.

### Changelog

- [x] `CHANGELOG.md` is updated with user-facing content relevant to this PR. If no changelog is required, then add the `no-changelog-required` label.

### Affected components
<!-- Remove non-affected components from the list -->

- [x] Server

### Server checklist
<!-- A checklist for server code -->

#### Catalog upgrade
<!-- Is hdb_catalog version bumped? -->
Does this PR change Hasura Catalog version?
- [x] No

#### Metadata
<!-- Hasura metadata changes -->

Does this PR add a new Metadata feature?
- [x] No

#### GraphQL
- [x] No new GraphQL schema is generated

#### Breaking changes

- [x] No Breaking changes

<!-- Explain briefly about your breaking changes below -->

codingkarthik reviewed (approved):
LGTM, Thank you 😄  . But on another note, is this the only piece of code that was lost in the rebase or were there other changes that might have been lost?

abooij commented:
> LGTM, Thank you 😄 . But on another note, is this the only piece of code that was lost in the rebase or were there other changes that might have been lost?

I have used the following git technique to figure this out. The commit that was rebased is be84310a4c9f4757575c8f59241d602c1c78ce6c, so we want to figure out all changes on `Schema.hs` since the last common commit of be84310a4c9f4757575c8f59241d602c1c78ce6c until right before the merging of #6116, which got merged as fd8d51a37abcf442353df1e5570b80941e1fb60a. Using the commands: (note the triple periods in the commit range!)

```shell
$ git diff be84310a4c9f4757575c8f59241d602c1c78ce6c...fd8d51a37abcf442353df1e5570b80941e1fb60a~ -- server/src-lib/Hasura/GraphQL/Schema.hs
$ git log be84310a4c9f4757575c8f59241d602c1c78ce6c...fd8d51a37abcf442353df1e5570b80941e1fb60a~ -- server/src-lib/Hasura/GraphQL/Schema.hs
```

one obtains all the changes not included in be84310a4c9f4757575c8f59241d602c1c78ce6c. The following commits are obtained:

* c6a03eabafb7eb93606937a0f81509c0612a23f8 (part of #5694)
  This introduced `safeSelectionSet`, and uses it throughout `Schema.hs`.
* 25773a7d7a6d7df7d1bd4ec675f6a52133d4d310 (part of #5694)
  Minor cleanup in how unauthenticated contexts are constructed
* 87e2f8234deb31b32f89b3e35158002be6e1f24a (part of #5797, not part of master, to be honest I'm not sure why this shows up in the git log)
* be84310a4c9f4757575c8f59241d602c1c78ce6c (part of #5797, not part of master, again I'm not sure why this shows up in the git log)
* 3ea611f9fd8f8b4425221f9ae299b51dce4bccec (part of #5938)
  Replaced `(RemoteSchemaInfo, G.Field G.NoFragments P.Variable)` with `RemoteField`.
* f46f8079425783578c90d1478200faf6f2801410 (part of #6017)
  Adjusts imports.
* 91d8a7ab614965cda5ed46513edbdb908860ce1a (part of #6017)
  Adjusts imports.
* 8d0afb6f92c6a2a3809c4aa96dab108d6cfd186b (part of #6003)
  Adds the `'Postgres` type parameter as required.
* a8ed6a82e209e507d131dc5ca5d24fb3c3a01c62 (part of #6053)
  Adjusts imports.
* 2cb08a89cbde087d86369d8d494f05e721cf4364 (part of #5719)
  Implemented by this PR.

The current version does not contain the phrases `RemoteSchemaInfo` or `selectionSet`. So from a functional point of view, this contains all the above changes.
